### PR TITLE
Fix for No mechanism for managing the switching sequence(Start/StopSt…

### LINKED ISSUE #3123: There is no mechanism for managing the switching sequence(Start/StopStreaming) between two navigation applications.
### Bug Report
There is no mechanism for managing the switching sequence(Start/StopStreaming) between two navigation applications.

##### Reproduction Steps
1. Register 2 apps:
・App1: Waze for iOS
・App2: Sygic for iOS
2. Activate App1
3. App1 starts video streaming
4. Activate App2
5. App1 is not displayed in the HMI
    It runs continuously in the background.

##### Expected Behavior
1. Activate App1: SDL->App1.OnHMIStatus(FULL)
2. App1 starts Video streaming
3. Activate App2: SDL->App2.OnHMIStatus(FULL)
4. Deactivate App1: SDL->App1.OnHMIStatus(BACKGROUND)
5. Receive EndService request from App1:SDL<-App1.EndServiceCallback(11)
6. Send StopSteaming notification of App1:HMI<-SDL:OnVideoDataStreaming(False)
7. Receive StartService request from App2:SDL<-App2.StartServiceCallback(11)
8. Send StarSteaming notification of App2:HMI<-SDL:OnVideoDataStreaming(True)
9. App2 starts video streaming

##### Observed Behavior
If App1.EndServiceCallback(11) arrives later than App2.StartServiceCallback(11)，
HMI failed to create socket for the video streaming of the App2,
because the socket of App1 hasn't been disconnected.

##### OS & Version Information
* SDL Core Version: 5.1.3
* Testing Against: SDL Core 5.1.3 + HMI running on our development environment

##### Test Case, Sample Code, and / or Example App
Logs:
[logInfoIOS_waze_Sygic.log](https://github.com/smartdevicelink/sdl_core/files/4152307/logInfoIOS_waze_Sygic.log)

##### Occurrence:
100%

##### Recovery Method:
None

### PR DESCRIPTION
**Pull request by zhouxin627**
Fixes #3123

This PR is **ready** for review.

### Risk
This PR makes **no** API changes.

### Testing Plan
Covered by Unit tests.

### Summary
**Reason**
1. Because switch between NaviApps. Only one app can be active at a time. Asynchronous notifications between apps may have timing issues.									
eg: App1.EndServiceCallback(11) arrives later than App2.StartServiceCallback(11)									
2. SDLCore does not effectively cover this case. Duplicate notifications affect the activated App.

**Solution**
1. When NaviApp2 activated and NaviApp1 hadn't stopNaviService, force stop NaviApp1 VPM as a failsafe process.
2. Double-check in StopNaviService to prevent NaviApp2 duplicate notifications

**Sequence**
image

### CLA
- [x] I have signed [the CLA](https://docs.google.com/forms/d/e/1FAIpQLSdsgJY33VByaX482zHzi-xUm49JNnmuJOyAM6uegPQ2LXYVfA/viewform)

### CODE REVIEW

**Comment by zhouxin627** [comment]
@theresalech please, be informed current PR is ready for review.

**Comment by iCollin** [comment]
Hello @zhouxin627,

It does not appear that you have signed [the CLA](https://docs.google.com/forms/d/e/1FAIpQLSdsgJY33VByaX482zHzi-xUm49JNnmuJOyAM6uegPQ2LXYVfA/viewform) which is required to contribute to the SmartDeviceLink library.

Please let me know when you've had a chance to sign,
Thank you

**Comment by zhouxin627** [comment]
> Hello @zhouxin627,
> 
> It does not appear that you have signed [the CLA](https://docs.google.com/forms/d/e/1FAIpQLSdsgJY33VByaX482zHzi-xUm49JNnmuJOyAM6uegPQ2LXYVfA/viewform) which is required to contribute to SmartDeviceLink libraries.
> 
> Please let me know when you've had a chance to sign,
> Thank you

Hi @iCollin ,
Thanks, I have signed the CLA.

File: src/components/application_manager/src/application_manager_impl.cc [hunk 1]
```diff
@@ -1825,6 +1825,30 @@ bool ApplicationManagerImpl::StartNaviService(
...
     }
 
+    {
+      /* Fix: For NaviApp1 Switch to NaviApp2, App1's Endcallback() arrives
+       later than App2's Startcallback(). Cause streaming issue on HMI.
+      */
+      sync_primitives::AutoLock lock(applications_list_lock_ptr_);
+      auto it_app = applications_.begin();
+      while (applications_.end() != it_app) {
+        ApplicationSharedPtr app = application((*it_app)->app_id());
```
[the remark this answers is not in the document]
Comment (zhouxin627) [reply src/components/application_manager/src/application_manager_impl.cc]: @iCollin Thanks, fixed in [b286dca](https://github.com/smartdevicelink/sdl_core/commit/b286dca84e1d866140302d7f3514e95c22810331)

File: src/components/application_manager/src/application_manager_impl.cc [hunk 2]
```diff
@@ -1825,6 +1825,30 @@ bool ApplicationManagerImpl::StartNaviService(
...
+        if (!app || (!app->is_navi() && !app->mobile_projection_enabled())) {
+          LOG4CXX_DEBUG(logger_,
+                        "Continue, Not Navi App Id: " << (*it_app)->app_id());
+          ++it_app;
+          continue;
+        } else {
```
[the remark this answers is not in the document]
Comment (zhouxin627) [reply src/components/application_manager/src/application_manager_impl.cc]: @iCollin Thanks, fixed in [b286dca](https://github.com/smartdevicelink/sdl_core/commit/b286dca84e1d866140302d7f3514e95c22810331)

**Pull request merged**